File: greedy/jumpGame.go
package greedy

var (
	counter = 0
)

func canJump(nums []int) bool {

	maxReach := 0
	currentVal := nums[0]
	nextIndex := 0
	// len = 1
	if len(nums) == 1 {
		return true
	}
	if currentVal == 0 && len(nums) != 1 {
		return false
	}
	for i := 1; i <= currentVal; i++ {
		if i+nums[i] > maxReach {
			maxReach = i + nums[i]
			nextIndex = i
		}
	}
	return canJump(nums[nextIndex:])
}

File: pythonSolution/dp/jumpGame.py
from typing import List

"""

Question 55:https://leetcode.com/problems/jump-game

"""


class Solution:

    # Generic DFS
    def canJumpDFS(self, nums: List[int]) -> bool:
        visited = set()
        nextIndex = [0]
        while nextIndex:
            cur = nextIndex.pop()
            if cur in visited:
                continue
            visited.add(cur)

            for i in range(1, nums[cur] + 1):
                next = i + cur
                if next >= len(nums):
                    break
                if next not in visited:
                    nextIndex.append(i + cur)

        return len(nums) - 1 in visited

    # Greedy approach with 1 checks
    def canJump(self, nums: List[int]) -> bool:
        lastGoodIndex = len(nums) - 1

        for i in range(lastGoodIndex - 1, -1, -1):
            if i + nums[i] >= lastGoodIndex:
                lastGoodIndex = i

        return lastGoodIndex == 0
